### PR TITLE
Bluetooth: Controller: Permit enabling already enabled advertising

### PR DESCRIPTION
**Pull request by cvinayak**
Permit enabling already enabled advertising. Enabling
advertising when it is already enabled can cause the
random address to change, as specified in the Bluetooth
Specifications.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/controller/Kconfig.ll_sw_split
Comment (sjanc) [inline subsys/bluetooth/controller/Kconfig.ll_sw_split:215]: Why add option that allows to break qualification?

File: subsys/bluetooth/controller/ll_sw/ull_adv.c
Comment (sjanc) [inline subsys/bluetooth/controller/ll_sw/ull_adv.c:664]: Shouldn't this (as per spec) also restart duration and events counter?

File: subsys/bluetooth/controller/Kconfig.ll_sw_split
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig.ll_sw_split:215]: This option had been the old behavior and does not break BT v4.2. The relaxed LE advertising enable was introduced in V5.0. Zephyr still keep the advertising state, and this option can be used by combined host+controller applications to valid their state machines.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c:664]: Good catch, will update with a fix.

File: subsys/bluetooth/controller/Kconfig.ll_sw_split
Comment (Thalley) [reply subsys/bluetooth/controller/Kconfig.ll_sw_split:215]: I agree with @sjanc here. 

If we can disable this, and thus not be BT 5 compliant, then we should at least set the reported BT version to that version, and alot disallow any BT5 and newer features. 

If we don't want to do that, I don't think we should allow this Kconfig option. 

There was another similar instance, where we actually, instead of removing it, added a warning if it was enabled, as it would break compliance and should not be used in a production build.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c
Comment (joerchan) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c:664]: The text says "can" not "shall". If you have a timer for this already you don't have to update the RPAs here.
The host wouldn't rely on something like this for behavior reason, since it cannot know how exactly the controller acts here.

However on the LE Set Extended Advertising Set Enable command it says:
"If the HCI_LE_Set_Extended_Advertising_Enable command is sent again for
an advertising set while that set is enabled, the timer used for the duration and
the number of events counter are reset and any change to the random address
shall take effect."

So any updates by the host for the advertising set random address should be updated at this point if not already.

File: subsys/bluetooth/controller/Kconfig.ll_sw_split
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig.ll_sw_split:215]: Just as the implementation can be changed by users to make the stack non-compliant, this Kconfig should not be seen any different. This Kconfig cannot be turned on by simply having it in the project conf file, it is only visible if advanced features of the controller is enabled explicitly. Even if enabled or disabled does not fail any latest HCI conformance test cases.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c
Comment (joerchan) [inline subsys/bluetooth/controller/ll_sw/ull_filter.c:593]: I have already suggested to remove this, but you probably meant to have reschedule here?

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c:593]: yes, it should be reschedule. The documentation for k_work has so many redirections as to the meaning of parameters to pass, its very easy to pick the wrong API.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c:664]: > The text says "can" not "shall". If you have a timer for this already you don't have to update the RPAs here.
The host wouldn't rely on something like this for behavior reason, since it cannot know how exactly the controller acts here.

"can" states a possibility and/or capability. And the possibility/capability to refresh address is a security feature in the implementation. Hence, will keep the RPA refresh if advertising is enabled on a already enabled case.

Comment (joerchan) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c:672]: I see this more as allowing the implementation to shorten the RPA timer if that is easier for the implementation. The security requirement is met as long as the timer is working correctly. Might be easier for implementations that always refresh the RPA of the advertiser when enabled (like the SoftDevice).
You don't force an RPA timeout when starting the disabled advertiser, so no reason to force it now IMO. Not a big deal though, but it would force RPA timeout for all users of the RPA timer (i.e all advertising sets and scanner/initiator). Which seems wasteful IMO.

Not a big deal though. Just a simplification you could do.

However applying the random address change should be enforced here.

File: subsys/bluetooth/controller/ll_sw/ull_filter.c
Comment (joerchan) [reply subsys/bluetooth/controller/ll_sw/ull_filter.c]: Yes, k_work is a bit more complicated now, so maybe just leave it. Using k_work is simplest when all you do is just having it at a periodic interval, which is the only requirement the RPA timer has.

File: subsys/bluetooth/controller/ll_sw/ull_adv.c
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c:672]: > You don't force an RPA timeout when starting the disabled advertiser, so no reason to force it now.

hmmm, yes, RPAs are only updated every `rpa_timeout_ms`. I will remove the update of all RPAs on enabling on already enabled advertiser.

> However applying the random address change should be enforced here.

Will work on this.

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c]: > applying the random address change should be enforced here

Done

File: tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c
Comment (Thalley) [inline tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:375]: I guess the number of sent expected is a function of the advertising interval and the timeout? 
I think this will be more clear if it was explicitly calculated, instead of implicitly:

 ```suggestion
	num_sent_expected = 4;
	ext_adv_param.timeout = 4 * ADV_INTERVAL;
```
or 
 ```suggestion
	ext_adv_param.timeout = 50;
	num_sent_expected = ext_adv_param.timeout / ADV_INTERVAL;
```

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_adv.c
Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ull_adv.c]: Done

File: tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c
Comment (cvinayak) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:375]: This is to be regarded as test vector input and it cannot be calculated due to fact that advertisement intervals are random and some timeout values (if ADV_INTERVAL changed by a developer/tester) influences the number of events.

Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:375]: If the interval is random and influences the number of events, how we can expect 4 advertisements to be sent?

Comment (cvinayak) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:375]: 4 events will be transmitted within 400 to 440 ms (4 * (BT_GAP_ADV_FAST_INT_MIN_1 + random_delay)) , the limit is set at 500 ms to ensure no more than 4 event is sent. Say events are changed to 11, then 11 events transmit in 1100 to 1210 ms, calculating using interval will give a limit of 1310 ms, which means there can be 11 or 12 events occurring based on random_delay applied on each event on air.

Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:375]: I see, so if the number of events increases then we might enter the possibility of have 1 more event than we expected, but that can be guarded with an __ASSERT of something to avoid that. 

But since you have already done the calculation on paper, why not bring it to the code, so that it becomes clear why we are waiting X ms for Y events? 

On another note: You mention that `4 * (BT_GAP_ADV_FAST_INT_MIN_1 + random_delay))` is 400-440ms. This implies that the random delay is 10ms, right? But `BT_GAP_ADV_FAST_INT_MIN_1` is 30 ms, so the above calculation doesn't make sense to me. Please elaborate.

Comment (cvinayak) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:375]: It is a bad idea to make tests and its input generic, and then add all sort of complexities. Tests being precise, deterministic and fail on any change is desired.

It was a typo, `BT_LE_ADV_CONN_NAME` uses `BT_GAP_ADV_FAST_INT_MIN_2`, 100 ms advertising interval.

Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:375]: I disagree; I don't think making it harder to update or change a test should be hard. 

I think making this change will not only make it possible to update the test at a later point if wanted, but also show the calculations that are currently implicitly made (and undocumented) with regards the number of events, the timeout values and the sleep values.

Comment (cvinayak) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:375]: I disagree too.

> update the test at a later point if wanted

That is recipe for disaster, and only be done if software requirements mandate so. I am aware that we have not documented and approved a test procedure here, as this is a test added (not updated any existing test case) as part of development to cover the implementation changes in this PR, which is the right thing to do in software process.

That said, you may send a PR with your proposed test implementation.

If there is no further change requests and do not agree to approve, you may dismiss your review in this PR.

File: subsys/bluetooth/controller/Kconfig.ll_sw_split
Comment (carlescufi) [reply subsys/bluetooth/controller/Kconfig.ll_sw_split:242]: @cvinayak could you please state what the purpose of this change is? Is it to allow users to qualify the controller under the 4.2 spec?

Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig.ll_sw_split:242]: 1. Qualification under v4.2 spec is valid under latest TCRL 2021-1, simple application like beacon can use it still, disabling privacy and filters reduce code size by 5KB.
2. Advertising enable has one parameter and we do not change (generate) private addresses on advertising enable, not necessary for different return codes, see @joerchan comment: https://github.com/zephyrproject-rtos/zephyr/pull/36446#discussion_r656155749 and https://github.com/zephyrproject-rtos/zephyr/pull/36446#discussion_r656220205
3. Zephyr Host does not call advertising enable if it is already enabled, this option compliments this if required (say to debug)

**Review by carlescufi — APPROVED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c
Comment (cvinayak) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c]: Added comments.

**Review by Thalley — APPROVED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c
Comment (ppryga) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c]: If I may.... I agree with @Thalley. 
Tests cases should be self explanatory. Magic numbers does not help at all. Also making tests hard to update/change is not a good way. Someone has to maintain it or change in future. Why not save effort and time to that person? 

If you verify some border conditions then you should  state it in e.g. test case name and then you can go with fixed values. Nevertheless that does not release you from requirement of documenting them (in comments or better by use of named macros). 

> That is recipe for disaster, and only be done if software requirements mandate so. 
Yes, tests are changed when the requirements are changed but why make it harder to do than it could be?

In my humble opinion there is more to do about this new block of code added to this looooooong function.

Comment (Thalley) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c]: The comments helps. I still stand by my original comment, but I not strongly enough that I will block a feature (or rather a fix to a compliance issue) because a test case may be improved for readability and refactorability.

Comment (kruithofa) [reply tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c]: FWIW I agree with Piotr and Emil

**Review by kruithofa — APPROVED** [review]
LGTM, but see comment regarding test

**Review by ppryga-nordic — APPROVED** [review]
I will be happy if follow up PR is immediately created to address topic from conversation about tests implementation.

**Comment by Thalley** [comment]
> I will be happy if follow up PR is immediately created to address topic from conversation about tests implementation.

@cvinayak correct me if I am wrong, but I don't think that is your intention based on https://github.com/zephyrproject-rtos/zephyr/pull/36446#discussion_r686806286, right?

**Comment by cvinayak** [comment]
> 
> 
> > I will be happy if follow up PR is immediately created to address topic from conversation about tests implementation.
> 
> @cvinayak correct me if I am wrong, but I don't think that is your intention based on [#36446 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/36446#discussion_r686806286), right?

Correct, I do not wish the change the test implementation to make it complicated such that if there is a bug in the test implementation, it would give false positives. The test cases itself should only be touched, when requirements change. Hiding expected test vector/values behind macros/defines is equally a maintenance overhead, that said, comments too have to be maintained if requirements change (I will not disagree to adding comments).

**Comment by ppryga** [comment]
> > > I will be happy if follow up PR is immediately created to address topic from conversation about tests implementation.

> @cvinayak correct me if I am wrong, but I don't think that is your intention based on #36446 (comment), right?

> Correct, I do not wish the change the test implementation to make it complicated such that if there is a bug in the test implementation, it would give false positives. The test cases itself should only be touched, when requirements change. Hiding expected test vector/values behind macros/defines is equally a maintenance overhead, that said, comments too have to be maintained if requirements change (I will not disagree to adding comments).

You are wrong @cvinayak. 

> Correct, I do not wish the change the test implementation to make it complicated such that if there is a bug in the test implementation, it would give false positives. The test cases itself should only be touched, when requirements change.

The test code is part of the complete solution (tested code + test code). Both parts should be written readable and understandable. Thanks to that it saves time for maintenance. The maintenance of test code is inevitable. Nobody says anything about chaining tests without change of requirements. But when requirements change then someone has to do that. Change requested by @Thalley saves time to do it by anyone.

> Hiding expected test vector/values behind macros/defines is equally a maintenance overhead, that said, comments too have to be maintained if requirements change (I will not disagree to adding comments).

Use of macros for constants is a maintenance cost, really? Use of macros etc. makes the code easier to understand, that saves a lot of time for maintainers. 

In my humble opinion **test code should be treated in the same way as regular code in regard of quality**. That is also production code. In TDD tests consume most of the time for implementation, so putting attention to quality of that saves time in total.

**Pull request merged**
